## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :x: [more...](#58c985a6-12b7-473e-adb4-e2b8b07f95a0) |
| mint-gateway-azure.sh | :x: [more...](#df277142-8e0c-41bf-af16-ca6277bee2c9) |
| mint-large-bucket.sh | :x: [more...](#401d255b-3437-4821-93f9-7369d9283a92) |
| mint-dist-xl.sh | :x: [more...](#98668aed-4d11-4dd5-b90a-cf50fe407a73) |
<a name="98668aed-4d11-4dd5-b90a-cf50fe407a73"></a>
### 6441-cddd415/mint-dist-xl.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32103
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 170613eb37b7:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... FAILED in 33 seconds
{
  "name": "awscli",
  "duration": 12617,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}

Executed 0 out of 13 tests successfully.
```
<a name="401d255b-3437-4821-93f9-7369d9283a92"></a>
### 6441-cddd415/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30322
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp dca3ac868a78:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... FAILED in 26 seconds
{
  "name": "awscli",
  "duration": 14604,
  "function": "make_bucket\n",
  "status": "FAIL",
  "error": "An error occurred (XMinioServerNotInitialized) when calling the CreateBucket operation (reached max retries: 4): Server not initialized, please try again."
}

Executed 0 out of 13 tests successfully.
```
<a name="58c985a6-12b7-473e-adb4-e2b8b07f95a0"></a>
### 6441-cddd415/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30417
ACCESS_KEY:           AKIAJ3ORETMQUQXBBWGA
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 0cd8eae13b95:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 3 minutes and 11 seconds
(2/13) Running aws-sdk-go tests ... done in 3 seconds
(3/13) Running aws-sdk-java tests ... done in 6 seconds
(4/13) Running aws-sdk-php tests ... done in 1 minutes and 21 seconds
(5/13) Running aws-sdk-ruby tests ... done in 56 seconds
(6/13) Running mc tests ... done in 8 minutes and 48 seconds
(7/13) Running minio-dotnet tests ... done in 8 minutes and 11 seconds
(8/13) Running minio-go tests ... done in 36 minutes and 18 seconds
(9/13) Running minio-java tests ... done in 18 minutes and 41 seconds
(10/13) Running minio-js tests ... done in 1 minutes and 34 seconds
(11/13) Running minio-py tests ... FAILED in 30 minutes and 6 seconds
{
  "status": "FAIL",
  "name": "minio-py:test_make_bucket_default_region",
  "duration": 1806019,
  "function": "make_bucket(bucket_name, location)",
  "args": {
    "location": "default value ('us-east-1')",
    "bucket_name": "minio-py-test-c3bbfd7f-5bd4-48f0-a6d8-4a9ec4001171"
  },
  "message": "HTTPConnectionPool(host='minikube', port=30417): Max retries exceeded with url: /minio-py-test-c3bbfd7f-5bd4-48f0-a6d8-4a9ec4001171/ (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 600, in urlopen\n    chunked=chunked)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 384, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 2, in raise_from\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 380, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.5/http/client.py\", line 1197, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.5/http/client.py\", line 297, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.5/http/client.py\", line 266, in _read_status\n    raise RemoteDisconnected(\"Remote end closed connection without\"\nhttp.client.RemoteDisconnected: Remote end closed connection without response\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 165, in test_make_bucket_default_region\n    client.make_bucket(bucket_name)\n  File \"/usr/local/lib/python3.5/dist-packages/minio/api.py\", line 286, in make_bucket\n    headers=headers)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/poolmanager.py\", line 322, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 667, in urlopen\n    **response_kw)\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/connectionpool.py\", line 638, in urlopen\n    _stacktrace=sys.exc_info()[2])\n  File \"/usr/local/lib/python3.5/dist-packages/urllib3/util/retry.py\", line 398, in increment\n    raise MaxRetryError(_pool, url, error or ResponseError(cause))\nurllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='minikube', port=30417): Max retries exceeded with url: /minio-py-test-c3bbfd7f-5bd4-48f0-a6d8-4a9ec4001171/ (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1707, in main\n    test_make_bucket_default_region(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 173, in test_make_bucket_default_region\n    raise Exception(err)\nException: HTTPConnectionPool(host='minikube', port=30417): Max retries exceeded with url: /minio-py-test-c3bbfd7f-5bd4-48f0-a6d8-4a9ec4001171/ (Caused by ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)))\n"
}

Executed 10 out of 13 tests successfully.
```
<a name="df277142-8e0c-41bf-af16-ca6277bee2c9"></a>
### 6441-cddd415/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:30807
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 40433c3c9450:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 3 minutes and 7 seconds
(2/13) Running aws-sdk-go tests ... done in 2 seconds
(3/13) Running aws-sdk-java tests ... done in 6 seconds
(4/13) Running aws-sdk-php tests ... done in 1 minutes and 18 seconds
(5/13) Running aws-sdk-ruby tests ... done in 41 seconds
(6/13) Running mc tests ... done in 7 minutes and 46 seconds
(7/13) Running minio-dotnet tests ... FAILED in 5 minutes and 21 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (Assert.AreEqual failed. Expected:<0x8D61716F54698E7-1>. Actual:<0x8D61716F4CCBEDF-1>. ) ---> Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException: Assert.AreEqual failed. Expected:<0x8D61716F54698E7-1>. Actual:<0x8D61716F4CCBEDF-1>. 
   at Microsoft.VisualStudio.TestTools.UnitTesting.Assert.HandleFail(String assertionName, String message, Object[] parameters)
   at Microsoft.VisualStudio.TestTools.UnitTesting.Assert.AreEqual[T](T expected, T actual, String message, Object[] parameters)
   at Minio.Functional.Tests.FunctionalTest.CopyObject_Test3(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 960
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 223

Executed 6 out of 13 tests successfully.
```

--- a/cmd/naughty-disk_test.go
+++ b/cmd/naughty-disk_test.go
@@ -52,6 +52,10 @@ func (d *naughtyDisk) IsOnline() bool {
 	return d.disk.IsOnline()
 }
 
+func (d *naughtyDisk) LastError() (err error) {
+	return nil
+}
+
 func (d *naughtyDisk) Close() (err error) {
 	if err = d.calcError(); err != nil {
 		return err

--- a/cmd/posix.go
+++ b/cmd/posix.go
@@ -274,6 +274,10 @@ func (s *posix) String() string {
 	return s.diskPath
 }
 
+func (s *posix) LastError() error {
+	return nil
+}
+
 func (s *posix) Close() error {
 	close(s.stopUsageCh)
 	s.connected = false

--- a/cmd/prepare-storage.go
+++ b/cmd/prepare-storage.go
@@ -120,12 +120,24 @@ var errXLV3ThisEmpty = fmt.Errorf("XL format version 3 has This field empty")
 // connect to list of endpoints and load all XL disk formats, validate the formats are correct
 // and are in quorum, if no formats are found attempt to initialize all of them for the first
 // time. additionally make sure to close all the disks used in this attempt.
-func connectLoadInitFormats(firstDisk bool, endpoints EndpointList, setCount, drivesPerSet int) (*formatXLV3, error) {
+func connectLoadInitFormats(retryCount int, firstDisk bool, endpoints EndpointList, setCount, drivesPerSet int) (*formatXLV3, error) {
+	// Initialize all storage disks
 	storageDisks, err := initStorageDisks(endpoints)
 	if err != nil {
 		return nil, err
 	}
 	defer closeStorageDisks(storageDisks)
+
+	// Connect to all storage disks, a connection failure will be
+	// only logged after some retries.
+	for _, disk := range storageDisks {
+		if disk != nil {
+			connectErr := disk.LastError()
+			if connectErr != nil && retryCount >= 5 {
+				logger.Info("Unable to connect to %s: %v\n", disk.String(), connectErr.Error())
+			}
+		}
+	}
 
 	// Attempt to load all `format.json` from all disks.
 	formatConfigs, sErrs := loadFormatXLAll(storageDisks)
@@ -238,8 +250,8 @@ func waitForFormatXL(ctx context.Context, firstDisk bool, endpoints EndpointList
 	retryTimerCh := newRetryTimerSimple(doneCh)
 	for {
 		select {
-		case _ = <-retryTimerCh:
-			format, err := connectLoadInitFormats(firstDisk, endpoints, setCount, disksPerSet)
+		case retryCount := <-retryTimerCh:
+			format, err := connectLoadInitFormats(retryCount, firstDisk, endpoints, setCount, disksPerSet)
 			if err != nil {
 				switch err {
 				case errNotFirstDisk:

--- a/cmd/storage-interface.go
+++ b/cmd/storage-interface.go
@@ -27,7 +27,9 @@ type StorageAPI interface {
 
 	// Storage operations.
 	IsOnline() bool // Returns true if disk is online.
+	LastError() error
 	Close() error
+
 	DiskInfo() (info DiskInfo, err error)
 
 	// Volume operations.

--- a/cmd/storage-rpc-client.go
+++ b/cmd/storage-rpc-client.go
@@ -104,6 +104,8 @@ func toStorageErr(err error) error {
 type StorageRPCClient struct {
 	*RPCClient
 	connected bool
+	// Plain error of the last RPC call
+	lastRPCError error
 }
 
 // Stringer provides a canonicalized representation of network device.
@@ -112,6 +114,11 @@ func (client *StorageRPCClient) String() string {
 	// Remove the storage RPC path prefix, internal paths are meaningless. why?
 	url.Path = strings.TrimPrefix(url.Path, storageServicePath)
 	return url.String()
+}
+
+// LastError - returns the last RPC call result, nil or error if any
+func (client *StorageRPCClient) LastError() error {
+	return client.lastRPCError
 }
 
 // Close - closes underneath RPC client.
@@ -125,14 +132,22 @@ func (client *StorageRPCClient) IsOnline() bool {
 	return client.connected
 }
 
+func (client *StorageRPCClient) connect() {
+	err := client.Call(storageServiceName+".Connect", &AuthArgs{}, &VoidReply{})
+	client.lastRPCError = err
+	client.connected = err == nil
+}
+
 func (client *StorageRPCClient) call(handler string, args interface {
 	SetAuthArgs(args AuthArgs)
 }, reply interface{}) error {
+
 	if !client.connected {
 		return errDiskNotFound
 	}
 
 	err := client.Call(handler, args, reply)
+	client.lastRPCError = err
 	if err == nil {
 		return nil
 	}
@@ -318,6 +333,7 @@ func newStorageRPC(endpoint Endpoint) *StorageRPCClient {
 	logger.FatalIf(err, "Unable to parse storage RPC Host", context.Background())
 	rpcClient, err := NewStorageRPCClient(host, endpoint.Path)
 	logger.FatalIf(err, "Unable to initialize storage RPC client", context.Background())
-	rpcClient.connected = rpcClient.Call(storageServiceName+".Connect", &AuthArgs{}, &VoidReply{}) == nil
+	// Attempt first try connection and save error if any.
+	rpcClient.connect()
 	return rpcClient
 }

